clang/lib/CodeGen/CGExprScalar.cpp

```diff
@@ -4139,6 +4139,56 @@ static bool isCheapEnoughToEvaluateUnconditionally(const Expr *E,
   // exist in the source-level program.
 }
 
+static Value *insertAddressSpaceCast(Value *V, unsigned NewAS) {
+  llvm::PointerType *VTy = cast<llvm::PointerType>(V->getType());
+  if (VTy->getAddressSpace() == NewAS) {
+    return V;
+  }
+  llvm::PointerType *VTyNewAS =
+      llvm::PointerType::get(VTy->getElementType(), NewAS);
+
+  if (auto *Constant = dyn_cast<llvm::Constant>(V)) {
+    return llvm::ConstantExpr::getAddrSpaceCast(Constant, VTyNewAS);
+  }
+
+  llvm::Instruction *NewV =
+      new llvm::AddrSpaceCastInst(V, VTyNewAS, V->getName() + ".ascast.rhs");
+  NewV->insertAfter(cast<llvm::Instruction>(V));
+  return NewV;
+}
+
+static void ensureSameAddrSpace(Value *&RHS, Value *&LHS,
+                                bool CanInsertAddrspaceCast,
+                                const LangOptions &Opts,
+                                const ASTContext &Context) {
+  if (RHS->getType() == LHS->getType()) {
+    return;
+  }
+
+  llvm::PointerType *RHSTy = dyn_cast<llvm::PointerType>(RHS->getType());
+  llvm::PointerType *LHSTy = dyn_cast<llvm::PointerType>(LHS->getType());
+  if (!RHSTy || !LHSTy ||
+      RHSTy->getAddressSpace() == LHSTy->getAddressSpace()) {
+    return;
+  }
+
+  if (!CanInsertAddrspaceCast)
+    // Pointers have different address spaces and we cannot do anything with
+    // this.
+    llvm_unreachable("Pointers are expected to have the same address space.");
+
+  // Language rules define if it is legal to cast from one address space to
+  // another, and which address space we should use as a "common denominator"
```

Comment:
I think we could add a comment that in case of SYCL this "common denominator" is OpenCL generic address space.